File: nb_crm_flutter/lib/features/org/domain/company_profile.dart
/// Shared company-profile fields used by Organization and Institute.
class CompanyProfileFields {
  const CompanyProfileFields({
    this.registrationNo,
    this.establishmentYear,
    this.contactPerson,
    this.mobileNo,
    this.contactNo,
    this.email,
    this.webAddress,
    this.panNo,
    this.gstNo,
    this.cinNo,
    this.country,
    this.state,
    this.city,
    this.address1,
    this.address2,
    this.pinCode,
    this.tagLine,
    this.hostingUrl,
    this.pageSize,
    this.dateFormat,
    this.timeZone,
    this.socialPostUrl,
    this.bankName,
    this.accountHolderName,
    this.bankAccountNo,
    this.ifscCode,
    this.bankBranch,
  });

  final String? registrationNo;
  final int? establishmentYear;
  final String? contactPerson;
  final String? mobileNo;
  final String? contactNo;
  final String? email;
  final String? webAddress;
  final String? panNo;
  final String? gstNo;
  final String? cinNo;
  final String? country;
  final String? state;
  final String? city;
  final String? address1;
  final String? address2;
  final String? pinCode;
  final String? tagLine;
  final String? hostingUrl;
  final String? pageSize;
  final String? dateFormat;
  final String? timeZone;
  final String? socialPostUrl;
  final String? bankName;
  final String? accountHolderName;
  final String? bankAccountNo;
  final String? ifscCode;
  final String? bankBranch;

  factory CompanyProfileFields.fromJson(Map<String, dynamic> json) {
    return CompanyProfileFields(
      registrationNo: json['registrationNo'] as String?,
      establishmentYear: _asIntOrNull(json['establishmentYear']),
      contactPerson: json['contactPerson'] as String?,
      mobileNo: json['mobileNo'] as String?,
      contactNo: json['contactNo'] as String?,
      email: json['email'] as String?,
      webAddress: json['webAddress'] as String?,
      panNo: json['panNo'] as String?,
      gstNo: json['gstNo'] as String?,
      cinNo: json['cinNo'] as String?,
      country: json['country'] as String?,
      state: json['state'] as String?,
      city: json['city'] as String?,
      address1: json['address1'] as String?,
      address2: json['address2'] as String?,
      pinCode: json['pinCode'] as String?,
      tagLine: json['tagLine'] as String?,
      hostingUrl: json['hostingUrl'] as String?,
      pageSize: json['pageSize'] as String?,
      dateFormat: json['dateFormat'] as String?,
      timeZone: json['timeZone'] as String?,
      socialPostUrl: json['socialPostUrl'] as String?,
      bankName: json['bankName'] as String?,
      accountHolderName: json['accountHolderName'] as String?,
      bankAccountNo: json['bankAccountNo'] as String?,
      ifscCode: json['ifscCode'] as String?,
      bankBranch: json['bankBranch'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'registrationNo': registrationNo,
        'establishmentYear': establishmentYear,
        'contactPerson': contactPerson,
        'mobileNo': mobileNo,
        'contactNo': contactNo,
        'email': email,
        'webAddress': webAddress,
        'panNo': panNo,
        'gstNo': gstNo,
        'cinNo': cinNo,
        'country': country,
        'state': state,
        'city': city,
        'address1': address1,
        'address2': address2,
        'pinCode': pinCode,
        'tagLine': tagLine,
        'hostingUrl': hostingUrl,
        'pageSize': pageSize,
        'dateFormat': dateFormat,
        'timeZone': timeZone,
        'socialPostUrl': socialPostUrl,
        'bankName': bankName,
        'accountHolderName': accountHolderName,
        'bankAccountNo': bankAccountNo,
        'ifscCode': ifscCode,
        'bankBranch': bankBranch,
      };

  static int? _asIntOrNull(Object? value) {
    if (value == null) return null;
    if (value is int) return value;
    if (value is num) return value.toInt();
    return int.tryParse(value.toString());
  }
}

class ParentOrganizationRef {
  const ParentOrganizationRef({
    required this.id,
    required this.code,
    required this.name,
  });

  final String id;
  final String code;
  final String name;

  factory ParentOrganizationRef.fromJson(Map<String, dynamic> json) {
    return ParentOrganizationRef(
      id: json['id']?.toString() ?? '',
      code: json['code'] as String? ?? '',
      name: json['name'] as String? ?? '',
    );
  }
}

File: backend/src/modules/organization/companyProfile.ts
/** Shared company-profile + bank fields for Organization and Institute. */

export type CompanyProfileInput = {
  registrationNo?: string | null;
  establishmentYear?: number | null;
  contactPerson?: string | null;
  mobileNo?: string | null;
  contactNo?: string | null;
  email?: string | null;
  webAddress?: string | null;
  panNo?: string | null;
  gstNo?: string | null;
  cinNo?: string | null;
  country?: string | null;
  state?: string | null;
  city?: string | null;
  address1?: string | null;
  address2?: string | null;
  pinCode?: string | null;
  tagLine?: string | null;
  hostingUrl?: string | null;
  pageSize?: string | null;
  dateFormat?: string | null;
  timeZone?: string | null;
  socialPostUrl?: string | null;
  bankName?: string | null;
  accountHolderName?: string | null;
  bankAccountNo?: string | null;
  ifscCode?: string | null;
  bankBranch?: string | null;
};

const OPTIONAL_STRING_KEYS = [
  'registrationNo',
  'contactPerson',
  'mobileNo',
  'contactNo',
  'email',
  'webAddress',
  'panNo',
  'gstNo',
  'cinNo',
  'country',
  'state',
  'city',
  'address1',
  'address2',
  'pinCode',
  'tagLine',
  'hostingUrl',
  'pageSize',
  'dateFormat',
  'timeZone',
  'socialPostUrl',
  'bankName',
  'accountHolderName',
  'bankAccountNo',
  'ifscCode',
  'bankBranch',
] as const;

export const REQUIRED_COMPANY_FIELDS: Array<{ key: keyof CompanyProfileInput; label: string }> = [
  { key: 'contactPerson', label: 'Contact Person' },
  { key: 'mobileNo', label: 'Mobile No.' },
  { key: 'panNo', label: 'PAN No.' },
  { key: 'gstNo', label: 'GST No.' },
  { key: 'email', label: 'Email' },
  { key: 'country', label: 'Country' },
  { key: 'state', label: 'State' },
  { key: 'city', label: 'City' },
  { key: 'address1', label: 'Address 1' },
  { key: 'pinCode', label: 'Pin Code' },
];

function trimOrNull(v: unknown): string | null {
  if (v == null) return null;
  const s = String(v).trim();
  return s.length ? s : null;
}

export function parseCompanyProfileFromBody(body: Record<string, unknown>): CompanyProfileInput {
  const out: CompanyProfileInput = {};
  for (const key of OPTIONAL_STRING_KEYS) {
    if (Object.prototype.hasOwnProperty.call(body, key)) {
      (out as Record<string, string | null>)[key] = trimOrNull(body[key]);
    }
  }
  if (Object.prototype.hasOwnProperty.call(body, 'establishmentYear')) {
    const raw = body.establishmentYear;
    if (raw === null || raw === '' || raw === undefined) {
      out.establishmentYear = null;
    } else {
      const n = Number(raw);
      if (!Number.isFinite(n) || n < 1800 || n > 2100) {
        throw new Error('Invalid establishment year');
      }
      out.establishmentYear = Math.trunc(n);
    }
  }
  return out;
}

export function assertRequiredCompanyFields(profile: CompanyProfileInput, opts?: { requireAll?: boolean }) {
  if (!opts?.requireAll) return;
  const missing: string[] = [];
  for (const { key, label } of REQUIRED_COMPANY_FIELDS) {
    const v = profile[key];
    if (v == null || String(v).trim() === '') missing.push(label);
  }
  if (missing.length) {
    throw new Error(`Missing required fields: ${missing.join(', ')}`);
  }
}

/** Build Prisma data object from parsed profile (only keys present). */
export function companyProfileToPrismaData(profile: CompanyProfileInput): Record<string, unknown> {
  const data: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(profile)) {
    if (v !== undefined) data[k] = v;
  }
  return data;
}
